near-sdk/src/private/abi.rs

```diff
@@ -141,28 +144,142 @@ pub struct AbiFunction {
+        /// JSON Subschema that represents this type (can be an inline primitive, a reference to the root schema and a few other corner-case things).
+        type_schema: Schema,
+    },
+    Borsh {
+        /// Inline Borsh schema that represents this type.
+        #[serde(with = "BorshSchemaContainerDef")]
+        type_schema: BorshSchemaContainer,
+    },
 }
 
-/// Represents how instances of a certain type are serialized in a certain context. Same type
-/// can have different serialization types associated with it depending on where they occur.
-#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
-#[serde(rename_all = "lowercase")]
-pub enum AbiSerializationType {
-    Json,
-    Borsh,
+// TODO: Maybe implement `Clone` for `BorshSchemaContainer` in borsh upstream?
+impl Clone for AbiType {
+    fn clone(&self) -> Self {
+        match self {
+            Self::Json { type_schema } => Self::Json { type_schema: type_schema.clone() },
+            Self::Borsh { type_schema } => {
+                let type_schema = BorshSchemaContainer {
+                    declaration: type_schema.declaration.clone(),
+                    definitions: type_schema
+                        .definitions
+                        .iter()
+                        .map(|(k, v)| (k.clone(), remote_serde_borsh::clone_definition(v)))
+                        .collect(),
+                };
+                Self::Borsh { type_schema }
+            }
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(remote = "BorshSchemaContainer")]
+struct BorshSchemaContainerDef {
+    declaration: Declaration,
+    #[serde(with = "remote_serde_borsh")]
+    definitions: HashMap<Declaration, Definition>,
+}
+
+/// This submodules follows https://serde.rs/remote-derive.html to derive Serialize/Deserialize for
+/// `BorshSchemaContainer` parameters. The top-level serialization type is `HashMap<Declaration, Definition>`
+/// for the sake of being easily plugged into `BorshSchemaContainerDef` (see its parameters).
+mod remote_serde_borsh {
+    use super::*;
+    use serde::ser::SerializeMap;
+    use serde::{Deserializer, Serializer};
+
+    #[derive(Serialize, Deserialize)]
+    #[serde(remote = "Fields")]
+    enum FieldsDef {
+        NamedFields(Vec<(FieldName, Declaration)>),
+        /// The struct with unnamed fields, structurally identical to a tuple.
+        UnnamedFields(Vec<Declaration>),
+        /// The struct with no fields.
+        Empty,
+    }
+
+    #[derive(Serialize, Deserialize)]
+    #[serde(remote = "Definition")]
+    enum DefinitionDef {
+        Array {
+            length: u32,
+            elements: Declaration,
+        },
+        Sequence {
+            elements: Declaration,
+        },
+        Tuple {
+            elements: Vec<Declaration>,
+        },
+        Enum {
+            variants: Vec<(VariantName, Declaration)>,
+        },
+        Struct {
+            #[serde(with = "FieldsDef")]
+            fields: Fields,
```

Comment:
The fields key feels redundant. Is there a reason this and others can't be inlined to avoid unnecessary data in abi?

Reply:
I am just following borsh-rs here: https://github.com/near/borsh-rs/blob/7325a0aab74049237b9f978e1e2c0fcf7bb799c2/borsh/src/schema.rs#L46

The structs need to be duplicated and match the other struct one for one so I can use serde's remote derivation (see https://serde.rs/remote-derive.html).